pilot/cmd/pilot-agent/options/options.go

```diff
@@ -89,7 +89,9 @@ var (
 	// DNSCaptureByAgent is a copy of the env var in the init code.
 	DNSCaptureByAgent = env.RegisterBoolVar("ISTIO_META_DNS_CAPTURE", false,
 		"If set to true, enable the capture of outgoing DNS packets on port 53, redirecting to istio-agent on :15053")
-
+	// InvalidDropByIptables is a copy of the env var in the init code.
+	InvalidDropByIptables = env.RegisterBoolVar("ISTIO_META_INVALID_DROP", false,
```

Comment:
I don't think we need to actually define it? We only need it to present on pod spec to let CNI consume it.

Reply:
@bianpengyuan CNI consumes it from here 
https://github.com/istio/istio/blob/master/tools/istio-iptables/pkg/capture/run.go#L360 
But the config was constructed here https://github.com/istio/istio/blob/master/tools/istio-iptables/pkg/cmd/root.go#L101 
That's why we need to set it from here https://github.com/istio/istio/pull/36667/files#diff-b59f1b2d634969faeda5096d06c7626e4de94d73c437391d08e0e91bbe9a1e21R57

Reply:
Not sure I follow. I think you just need the env variable name at CNI when it checks env from pod? Can you just use the existing one defined at https://github.com/istio/istio/blob/master/tools/istio-iptables/pkg/cmd/root.go#L43? The agent does not really have any logic based on this env variable, so feels not right to define it here.

Reply:
@bianpengyuan  Actually this PR is quite similar with this one https://github.com/istio/istio/pull/32416/files.
I went through the code again and this is the current flow:

1. Istio CNI redirect. go will read the `ISTIO_META_INVALID_DROP` from pod env variable and assign to `rdrct.invalidDrop`
2. Istio CNI iptables.go read this flag from `#1` `rdrct.invalidDrop` and set to `viper.Set(constants.DropInvalid, rdrct.invalidDrop)`
3. Istio CNI iptables.go will program the iptables rules based on the parameters provided in Redirect, it will get the iptables command from Istio tools root.go. While the root.go will `constructConfig()` based on the `viper.GetBool(constants.DropInvalid)` and the value was set in `#2`.
4. Istio CNI generates the iptables with DropInvalid = true, so the invalid drop policy is there.

If my understanding is correct, your concern is in `#3`, the Istio CNI can directly read the flag from [here](https://github.com/istio/istio/blob/master/tools/istio-iptables/pkg/cmd/root.go#L43) , that would possibly work but it will not follow the `viper` style.

Reply:
No, the flow is right. All I mean is that we do not need to define it at agent. At `#1`, after adding the env var via PROXY_METADATA, it will show up at istio proxy env var list. CNI can parse that env var even without actually defining it at agent. Not sure if I miss something here?

Reply:
@bianpengyuan  Just tested again without this PR Istio CNI will always get that env as false even specified it as true in the proxy container spec and in the printenv, so the invalid drop rule is not there:
```
kubectl exec  -n httpbin-pre-production swissknife-test-1 -c istio-proxy  -- printenv |  grep ISTIO_META_INVALID_DROP
ISTIO_META_INVALID_DROP=true
```
So seems like the CNI can't parse env var here and more precisely, the istio-iptables binary can't read it from the env since istio-cni calls istio-iptables with params to setup pod netns.

So maybe I can refactor this PR to remove the ISTIO_META_INVALID_DROP in CNI as below but keep it in agent:
https://github.com/istio/istio/blob/master/tools/istio-iptables/pkg/cmd/root.go#L43

Reply:
Yeah I understand we need this PR. I meant you don't need to define it again in the agent code (i.e. revert change at this options.go file and use the init container env var that you already have to get the name of invalid_drop env var.) CNI does not read the env var at proxy container runtime. It reads from the pod spec directly (i.e. cni is doing something like `kubectl get pod -oyaml | grep ISTIO_META_INVALID_DROP` before pod starts up). If you keep this all the changes you have in this PR and revert this options.go file. Things should still work.

Reply:
And I think the reason why `istio-iptables` can't read the `ISTIO_META_INVALID_DROP` from proxy container env var is that in CNI mode this binary runs in host network and then do nsenter and inject iptables to pod network ns, so it can't read the container env var. 
https://github.com/istio/istio/blob/master/tools/istio-iptables/pkg/cmd/root.go#L331
In other words, CNI can only set the default value and need agent to get it from the container env var in spec and pass to CNI.

Reply:
Agent does not pass env var to CNI. All of CNI logic executes before pod starts up.

Reply:
>CNI can only set the default value and need agent to get it from the container env var in spec and pass to CNI.

It's actually istio-iptables can only set the default value and need CNI to get it from the container env var and pass to the istio-iptables with the addition of an CLI argument.

It reminds me of another upstream issue that the smart dns proxy was not working in CNI mode that's why there was a fix https://github.com/istio/istio/pull/30831.  It's similar issue there since with CNI the iptables binary runs on the node, it's getting the env var from the node, not from the container. With the addition of an istio-iptables CLI argument, CNI can read it from the pod info and pass to istio-iptables.

I think I'm clear now. There are two flows need to take care, one is the sidecar injection through init-container and another is the CNI and this PR is particular for the CNI flow.

Thanks @bianpengyuan  for you comments, I just refactor the PR.

Reply:
@bianpengyuan  I've tested this PR and it's also working, please let me know if you have any comments, thanks!